pkg/agent/controller/serviceimport.go

```diff
@@ -0,0 +1,217 @@
+
+					if !ok {
+						klog.Errorf("Failed to convert deleted tombstone object %v  to serviceimport", tombstone.Obj)
+						return
+					}
+				}
+				if si.Spec.Type != lighthousev2a1.Headless {
+					return
+				}
+				c.serviceImportDeletedMap.Store(key, si)
+				c.queue.AddRateLimited(key)
+			}
+		},
+	})
+
+	go c.serviceInformer.Run(stopCh)
+	go c.runServiceImportWorker()
+
+	go func(stopCh <-chan struct{}) {
+		<-stopCh
+		c.queue.ShutDown()
+
+		klog.Infof("ServiceImport Controller stopped")
+	}(stopCh)
+
+	return nil
+}
+
+func (c *ServiceImportController) runServiceImportWorker() {
+	for {
+		keyObj, shutdown := c.queue.Get()
+		if shutdown {
+			klog.Infof("Lighthouse watcher for ServiceImports stopped")
+			return
+		}
+
+		key := keyObj.(string)
+
+		func() {
+			defer c.queue.Done(key)
+			obj, exists, err := c.serviceInformer.GetIndexer().GetByKey(key)
+
+			if err != nil {
+				klog.Errorf("Error retrieving the object with store is  %v from the cache: %v", c.serviceInformer.GetIndexer().ListKeys(), err)
+				// requeue the item to work on later
+				c.queue.AddRateLimited(key)
+
+				return
+			}
+
+			c.queue.Forget(key)
+
+			if exists {
+				c.serviceImportCreatedOrUpdated(obj, key)
+			} else {
+				c.serviceImportDeleted(key)
+			}
+		}()
+	}
+}
+
+func (c *ServiceImportController) serviceImportCreatedOrUpdated(obj interface{}, key string) {
+	if _, found := c.endpointControllers.Load(key); found {
+		klog.V(log.DEBUG).Infof("The endpoint controller is already running fof %q", key)
+		return
+	}
+
+	serviceImportCreated := obj.(*lighthousev2a1.ServiceImport)
+	if serviceImportCreated.Spec.Type != lighthousev2a1.Headless {
+		return
+	}
+
+	annotations := serviceImportCreated.ObjectMeta.Annotations
+	serviceNameSpace := annotations[originNamespace]
+	serviceName := annotations[originName]
+	var service *corev1.Service
+
+	service, err := c.kubeClientSet.CoreV1().Services(serviceNameSpace).Get(serviceName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return
+		}
+
+		c.queue.AddRateLimited(key)
+		klog.Errorf("Error retrieving the service  %q from the namespace %q : %v", serviceName, serviceNameSpace, err)
+
+		return
+	}
+
+	if service.Spec.Selector == nil {
+		klog.Errorf("The service %s/%s without a Selector is not supported", serviceNameSpace, serviceName)
+		return
+	}
+
+	labelSelector := labels.Set(service.Spec.Selector).AsSelector()
+	endpointController, err := NewEndpointController(c.kubeClientSet, serviceImportCreated.ObjectMeta.UID,
+		serviceImportCreated.ObjectMeta.Name, c.clusterID)
+
+	if err != nil {
+		klog.Errorf("Error creating Endpoint controller for service %s/%s: %v", serviceNameSpace, serviceName, err)
+		return
+	}
+
+	err = endpointController.Start(endpointController.stopCh, labelSelector)
+	if err != nil {
+		klog.Errorf("Error starting Endpoint controller for service %s/%s: %v", serviceNameSpace, serviceName, err)
+		return
+	}
```

Comment:
So if `NewEndpointController` or `Start` return an error should we re-queue? However, they don't actually have code paths that return an error. So I suggest simplifying this and change the signatures to not return an error.

Reply:
Done.